Firestore/core/src/firebase/firestore/remote/grpc_stream.h

```diff
@@ -0,0 +1,261 @@
+   */
+  bool WriteAndFinish(grpc::ByteBuffer&& message);
+
+  bool IsFinished() const { return state_ == State::Finished; }
+
+  /**
+   * Returns the metadata received from the server. It is only valid to call
+   * this method once the stream has opened.
+   */
+  MetadataT GetResponseHeaders() const;
+
+  // These are part of `GrpcStream` implementation details that are only public
+  // for the sake of simplicity; do not use.
+  void OnStart();
+  void OnRead(const grpc::ByteBuffer& message);
+  void OnWrite();
+  void OnOperationFailed();
+  void OnFinishedByServer(const grpc::Status& status);
+  void OnFinishedByClient();
+  void RemoveOperation(const StreamOperation* to_remove);
+
+ private:
+  void Read();
+  StreamWrite* BufferedWrite(grpc::ByteBuffer&& message);
+
+  // A blocking function that waits until all the operations issued by this
+  // stream come out from the gRPC completion queue. Once they do, it is safe to
+  // delete this `GrpcStream` (thus releasing `grpc::ClientContext`). This
+  // function should only be called during the stream finish.
+  //
+  // Important: before calling this function, the caller must be sure that any
+  // pending operations on the gRPC completion queue will come back quickly
+  // (either because the call has failed, or because the call has been
+  // canceled). Otherwise, this function will block indefinitely.
+  void FastFinishOperationsBlocking();
+
+  // Whether this stream belongs to the same generation as the observer.
+  bool SameGeneration() const;
+
+  // Creates and immediately executes an operation, storing a raw pointer to the
+  // operation.
+  template <typename Op, typename... Args>
+  void Execute(Args... args) {
+    operations_.push_back(StreamOperation::ExecuteOperation<Op>(
+        this, call_.get(), firestore_queue_, args...));
+  }
+
+  // The gRPC objects that have to be valid until the last gRPC operation
+  // associated with this call finishes. Note that `grpc::ClientContext` is
+  // _not_ reference-counted.
+  //
+  // Important: `call_` has to be destroyed before `context_`, so declaration
+  // order matters here. Despite the unique pointer, `call_` is actually
+  // a non-owning handle, and the memory it refers to (part of a gRPC memory
+  // arena) will be released once `context_` (which is owning) is released.
+  std::unique_ptr<grpc::ClientContext> context_;
+  std::unique_ptr<grpc::GenericClientAsyncReaderWriter> call_;
+
+  util::AsyncQueue* firestore_queue_ = nullptr;
+
+  GrpcStreamObserver* observer_ = nullptr;
+  int generation_ = -1;
+  internal::BufferedWriter buffered_writer_;
+
+  std::vector<StreamOperation*> operations_;
+
+  // The order of stream states is linear: a stream can never transition to an
+  // "earlier" state, only to a "later" one (e.g., stream can go from `Starting`
+  // to `Open`, but not vice versa). Intermediate states can be skipped (e.g.,
+  // a stream can go from `Starting` directly to `Finishing`).
+  enum class State {
```

Comment:
I'm surprised that this state is different from `FSTStreamState`. I'm worried that it's going to be confusing to talk about this implementation (e.g. "The stream state was open." -> "Did you mean GrpcStream.State or the portable stream state?")

I'm also concerned that we've implemented essentially a wrapper on top of a wrapper that we'll use to ultimately implement the portable stream API. I worry that this intermediate abstraction layer may not actually pay for itself.

Now that we shut down streams synchronously, how much would it hurt to make these reusable and just directly implement the higher-level portable stream abstraction directly here?